Update front end hover to include overlap stats

andyzorigin opened the pull request:
This PR updates front end hover to include overlap stats

This includes: 
1) Get the scenario_specs associated with a group, as overlap stats are at scenario_spec level, but the web displays at group level, and scenario_specs don't contain group info.
2) Add a script to aggregate the overlap stats at a group level.
3) makes changes to summarize.py to display group overlap stats for various models on front end.

Example run:
`python3 src/helm/benchmark/data_overlap/get_group_scenario_specs.py --run-specs src/helm/benchmark/presentation/run_specs.conf --output-data group_scenario_specs`

`
python3 get_group_overlap_stats.py --overlap_stats_path output_stats_pile_all --group_scenario_specs_path group_scenario_specs --output-data pile_group_overlap_stats.json --models "together/bloom" "together/gpt-j-6b" "together/gpt-neox-20b" "together/opt-66b" "together/opt-175b" "microsoft/TNLGv2_7B" "microsoft/TNLGv2_530B" "anthropic/stanford-online-all-v4-s3" "anthropic/claude-v1.3" "anthropic/claude-instant-v1" "together/yalm"
`

Example output:
```
{"group": "boolq", "scenario_specs": [{"class_name": "helm.benchmark.scenarios.boolq_scenario.BoolQScenario", "args": {"only_contrast": false}}, {"class_name": "helm.benchmark.scenarios.boolq_scenario.BoolQScenario", "args": {"only_contrast": "True"}}]}
{"group": "narrative_qa", "scenario_specs": [{"class_name": "helm.benchmark.scenarios.narrativeqa_scenario.NarrativeQAScenario", "args": {}}]}
{
...
```

```
{{"models": ["together/gpt-j-6b"], "group_overlap_stats_list": [{"group": "boolq", "num_instances": 3334, "num_overlapping_inputs": 1769, "num_overlapping_references": 0}, {"group": "narrative_qa", "num_instances": 355, "num_overlapping_inputs": 175, "num_overlapping_references": 0}, 
...
```

percyliang commented on scripts/data_overlap/get_group_overlap_stats.py:
```diff
@@ -0,0 +1,91 @@
+from light_scenario import GroupScenarioSpecs, GroupOverlapStats, AllGroupOverlapStats
+
+
+def save_group_overlap_stats_to_jsonl(all_group_overlap_stats: AllGroupOverlapStats, filename: str):
```
make method name reflect that we're appending to the file (and add docstring)

percyliang commented on scripts/data_overlap/get_group_overlap_stats.py:
```diff
@@ -0,0 +1,91 @@
+        n = data_overlap_stats_key.overlap_protocol_spec.n
+        num_overlapping_inputs = len(data_overlap_stats.instance_ids_with_overlapping_input)
+        num_overlapping_references = len(data_overlap_stats.instance_ids_with_overlapping_reference)
+        if n == 13:
```
Is there a way to avoid hardcoding?  At least add a comment for why this exists

andyzorigin replied:
Added as optional arg default to 13

percyliang commented on scripts/data_overlap/get_group_overlap_stats.py:
```diff
@@ -0,0 +1,91 @@
+
+    group_scenario_specs_jsons = open(args.group_scenario_specs_path, "r").readlines()
+
+    group_scenario_specs_list: List = []
```
Add element type for `List`

percyliang commented on scripts/data_overlap/get_group_overlap_stats.py:
```diff
@@ -0,0 +1,91 @@
+                group_num_instances += num_instances
+                group_num_overlapping_inputs += num_overlapping_inputs
+                group_num_overlapping_references += num_overlapping_references
+            except Exception:
```
In general, bad style to catch a blanket Exception - can easily result in silent errors...  I don't think there should be a try/Except here...

percyliang commented on scripts/data_overlap/get_group_overlap_stats.py:
```diff
@@ -0,0 +1,91 @@
+
+
+if __name__ == "__main__":
+    parser = argparse.ArgumentParser()
```
In general, could add more documentation on what this file is trying to do

andyzorigin replied:
Added

percyliang commented on scripts/data_overlap/get_group_overlap_stats.py:
```diff
@@ -0,0 +1,91 @@
+
+if __name__ == "__main__":
+    parser = argparse.ArgumentParser()
+    parser.add_argument("--overlap_stats_path", type=str, required=True, help="The path to the overlap stats file")
```
use kebab case consistently everywhere

Just call it `--overlap-stats` (since path is explicit) or at least make it consistent with `--output-data` (which does not have `path`).

percyliang commented on scripts/data_overlap/get_group_overlap_stats.py:
```diff
@@ -0,0 +1,91 @@
+    parser = argparse.ArgumentParser()
+    parser.add_argument("--overlap_stats_path", type=str, required=True, help="The path to the overlap stats file")
+    parser.add_argument(
+        "--group_scenario_specs_path", type=str, required=True, help="The path to the group scenario specs file"
```
See comment above

percyliang commented on scripts/data_overlap/get_group_overlap_stats.py:
```diff
@@ -0,0 +1,91 @@
+    args = parser.parse_args()
+
+    try:
+        os.remove(args.output_data)
```
Can we write this as: if exists, then remove

percyliang commented on src/helm/benchmark/data_overlap/get_group_scenario_specs.py:
```diff
@@ -0,0 +1,59 @@
+    run_entries = read_run_entries(args.run_specs).entries
+    run_specs = run_entries_to_run_specs(
+        run_entries=run_entries,
+        priority=4,
```
why is 4 hardcoded here?

andyzorigin replied:
We do this to use the same run_conf as the main helm repo to filter those that are not relevant/are problematic to data overlap (e.g. lextreme). I've changed to constant instead so at least it's not a magic number

percyliang commented on src/helm/benchmark/data_overlap/read_overlap_stats.py:
```diff
@@ -0,0 +1,37 @@
+            overlap_stats_dict = json.loads(overlap_stats_json)
+            overlap_stats_list.append(cattrs.structure(overlap_stats_dict, AllGroupOverlapStats))
+        return overlap_stats_list
+    except Exception:
```
Don't catch bare `Exception`

percyliang commented on src/helm/benchmark/static/overlap_stats.jsonl:
```diff
@@ -0,0 +1 @@
+{"models": ["together/bloom", "together/gpt-j-6b", "together/gpt-neox-20b", "together/opt-66b", "together/opt-175b", "microsoft/TNLGv2_7B", "microsoft/TNLGv2_530B", "anthropic/stanford-online-all-v4-s3", "anthropic/claude-v1.3", "anthropic/claude-instant-v1", "together/yalm"], "group_overlap_stats_list": [{"group": "boolq", "num_instances": 3334, "num_overlapping_inputs": 1769, "num_overlapping_r ...
```
Don't check this in.

percyliang commented on scripts/data_overlap/light_scenario.py:
```diff
@@ -7,6 +7,39 @@
+@dataclass(frozen=True)
+class GroupScenarioSpecs:
+    """
+    Scenario Specs of Group
```
Add a better docstring

yifanmai commented:
As discussed elsewhere:

- The workflow here is too complicated. The current workflow is: given the overlap output, run a script in HELM to get the groups, run another script to join the groups with the result, and then run summarize with that output. I want to just put the overlap output file in a subdirectory in the suite directory and then run summarize.
- `summarize.py` already has the mapping between `ScenarioSpec`s and groups, so we should use that rather than trying to recompute it from the run entries file.
- Philosophically, I would prefer if the overlap code lives in an entirely separate repo / package and depends on HELM. But we can punt on this.

andyzorigin commented:
Updated and simplified logic to summarize.py. Introduce a new yaml of output_stat filenames -> models

Files for the pile include metadata.yaml and output_stats_pile_all in codalab folder
https://worksheets.codalab.org/bundles/0xd66a96f625c84b16a66a885dca5c1fee

yifanmai commented on src/helm/benchmark/presentation/summarize.py:
```diff
@@ -315,6 +319,133 @@ def read_runs(self):
+                - model3
+
+            """
+            metadata_file_path: str = os.path.join(data_overlap_dir, "metadata.yaml")
```
Nvm.

andyzorigin replied:
Discussed how it's in data_overlap directory so it's fine

andyzorigin commented on src/helm/benchmark/presentation/summarize.py:
```diff
@@ -315,6 +319,133 @@ def read_runs(self):
+        Concretely:
+            - get group -> scenario_spec information from self.runs
+                run_spec data
+            - read the files in the data_overlap directory in run_suit_path
```
Updated

andyzorigin commented on src/helm/benchmark/data_overlap/data_overlap_spec.py:
```diff
@@ -0,0 +1,86 @@
+
+
+@dataclass(frozen=True)
+class OverlapProtocolSpec:
```
We haven't introduced it yet; this is identical with the other data_overlap_spec file

andyzorigin commented on src/helm/benchmark/presentation/summarize.py:
```diff
@@ -315,6 +319,133 @@ def read_runs(self):
+            - get aggregate stats for group -> overlap ratio
+        """
+
+        def get_group_to_scenario_specs(run_specs: List[RunSpec]) -> Dict[RunSpec, ScenarioSpec]:
```
Would it? I'd rather have a few nested functions that abstract this logic out rather than a huge function that needs to be read line by line, especially when read using an editor which can jump to the functions. It'll also be easier to share this logic if the need arises as is

andyzorigin commented on src/helm/benchmark/presentation/summarize.py:
```diff
@@ -315,6 +319,133 @@ def read_runs(self):
+                raise ValueError("Invalid YAML structure: 'file_models_mapping' key not found.")
+
+            file_metadata = {}
+            for entry in data["file_models_mapping"]:
```
What is "this"? I'm not convinced there's much value in adding a dataclass here. For `entry`, we quickly modify it since it's relative paths to construct a separate dict. And we do not use this file/data format anywhere else so there's no interoperability concerns.

andyzorigin replied:
Sure I'm happy to just remove that validation and assume the file format is structured correctly. We aren't assuming an adversarial/open setting for the file and the complexity/overhead to do more given that this file is small, used once, and hardly changes doesn't seem worth it. If things ever change, it won't be hard to add validation or a dataclass when needed

andyzorigin commented on src/helm/benchmark/presentation/summarize.py:
```diff
@@ -315,6 +319,133 @@ def read_runs(self):
+            for model_name in model_names:
+                # Assume model name will only be associated with single group overlap list for now
+                # can update to join lists if need arises
+                self.model_group_overlap_stats[model_name] = group_overlap_stats_list
```
Updated

yifanmai commented on src/helm/benchmark/presentation/summarize.py:
```diff
@@ -315,6 +319,128 @@ def read_runs(self):
+        """
+
+        def get_group_to_scenario_specs(run_specs: List[RunSpec]) -> Dict[str, ScenarioSpec]:
+            scenario_specs_to_groups: Dict = {}
```
Add type parameters to type annotation.

yifanmai commented on src/helm/benchmark/presentation/summarize.py:
```diff
@@ -315,6 +319,128 @@ def read_runs(self):
+                ):
+                    scenario_specs_to_groups[scenario_spec] = groups
+
+            group_to_scenario_specs: Dict = {}
```
Add type parameters to type annotation.
